@here/harp-mapview/lib/PolarTileDataSource.ts

```diff
@@ -0,0 +1,317 @@
+    private m_geometryLevelOffset: number;
+
+    private m_styleSetEvaluator?: StyleSetEvaluator;
+    private m_northPoleMaterial?: THREE.Material;
+    private m_southPoleMaterial?: THREE.Material;
+
+    constructor({
+        name = "polar",
+        styleSetName,
+        minZoomLevel,
+        maxZoomLevel,
+        // storageLevelOffset = -2,
+        storageLevelOffset = +1,
+        geometryLevelOffset = -1
+    }: {
+        name?: string;
+        styleSetName?: string;
+        minZoomLevel?: number;
+        maxZoomLevel?: number;
+        storageLevelOffset?: number;
+        geometryLevelOffset?: number;
+    }) {
+        super(name, styleSetName, minZoomLevel, maxZoomLevel, storageLevelOffset);
+
+        this.m_geometryLevelOffset = geometryLevelOffset;
+        this.cacheable = false;
+    }
+
+    dispose() {
+        if (this.m_northPoleMaterial) {
+            this.m_northPoleMaterial.dispose();
+            delete this.m_northPoleMaterial;
+        }
+        if (this.m_southPoleMaterial) {
+            this.m_southPoleMaterial.dispose();
+            delete this.m_southPoleMaterial;
+        }
+        if (this.m_styleSetEvaluator) {
+            delete this.m_styleSetEvaluator;
+        }
+    }
+
+    setStyleSet(styleSet?: StyleSet, languages?: string[]): void {
+        this.dispose();
+
+        if (styleSet !== undefined) {
+            this.m_styleSetEvaluator = new StyleSetEvaluator(styleSet);
+
+            const northEnv = new MapEnv({
+                $geometryType: "polygon",
+                $layer: "earth",
+                kind: "north_pole"
+            });
+
+            const northTechnique = this.m_styleSetEvaluator.getMatchingTechniques(northEnv);
+            if (northTechnique.length !== 0) {
+                this.m_northPoleMaterial = createMaterial({
+                    technique: northTechnique[0],
+                    level: 1
+                });
+            }
+
+            const southEnv = new MapEnv({
+                $geometryType: "polygon",
+                $layer: "earth",
+                kind: "south_pole"
+            });
+
+            const southTechnique = this.m_styleSetEvaluator.getMatchingTechniques(southEnv);
+            if (southTechnique.length !== 0) {
+                this.m_southPoleMaterial = createMaterial({
+                    technique: southTechnique[0],
+                    level: 1
+                });
+            }
+        }
+
+        this.mapView.markTilesDirty(this);
+    }
+
+    shouldRender(zoomLevel: number, tileKey: TileKey): boolean {
+        if (zoomLevel !== tileKey.level || tileKey.level < 2) {
+            return false;
+        }
+
+        const { north, south } = this.m_tilingScheme.getGeoBox(tileKey);
+
+        const shouldRender = north > this.m_maxLatitude || south < -this.m_maxLatitude;
+
+        return shouldRender;
+    }
+
+    shouldSubdivide(zoomLevel: number, tileKey: TileKey): boolean {
+        if (zoomLevel < tileKey.level) {
```

Comment:
<= ?

Reply:
nope, if the datasource has not shouldRender redefined, the tile will render at 
```zoomLevel === tileKey.level```.
So, why check sub-tiles if parent tile will be renderd already?

Reply:
that's why shouldSubdive should also return false when `zoomLevel === tileKey.level`

Reply:
Wow, you're right. I was thinking about initial method in DataSource where it does return true, and here it returns false. Good catch, thanks!